### PR TITLE
MULE-17396: JMS connection resources should be closed in a separate thread

### PR DESCRIPTION
**Pull request by plbalbi**

### CODE REVIEW

File: transports/jms/src/main/java/org/mule/transport/jms/DeferredJmsResourceCloser.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        super("DeferredJMSResourcesCloser");
+        this.connector = jmsConnector;
+        this.queue = deferredCloseQueue;
+    }
+
+
+    @Override
+    public void run()
+    {
+        // TODO: Change this logger to debug level
```
Comment (fsgonz) [inline transports/jms/src/main/java/org/mule/transport/jms/DeferredJmsResourceCloser.java]: change this to debug.

File: transports/jms/src/main/java/org/mule/transport/jms/JmsReplyToHandler.java [hunk 2]
```diff
@@ -118,9 +123,9 @@ public void processReplyTo(MuleEvent event, MuleMessage returnMessage, Object re
 
             // QoS support
             MuleMessage eventMsg = event.getMessage();
-            String ttlString = (String)eventMsg.getOutboundProperty(JmsConstants.TIME_TO_LIVE_PROPERTY);
-            String priorityString = (String)eventMsg.getOutboundProperty(JmsConstants.PRIORITY_PROPERTY);
-            String persistentDeliveryString = (String)eventMsg.getOutboundProperty(JmsConstants.PERSISTENT_DELIVERY_PROPERTY);
+            String ttlString = (String) eventMsg.getOutboundProperty(JmsConstants.TIME_TO_LIVE_PROPERTY);
```
Comment (fsgonz) [inline transports/jms/src/main/java/org/mule/transport/jms/JmsReplyToHandler.java]: static import

File: transports/jms/src/main/java/org/mule/transport/jms/JmsReplyToHandler.java [hunk 3]
```diff
@@ -118,9 +123,9 @@ public void processReplyTo(MuleEvent event, MuleMessage returnMessage, Object re
...
+            String priorityString = (String) eventMsg.getOutboundProperty(JmsConstants.PRIORITY_PROPERTY);
```
Comment (fsgonz) [inline transports/jms/src/main/java/org/mule/transport/jms/JmsReplyToHandler.java]: static import

File: transports/jms/src/main/java/org/mule/transport/jms/JmsReplyToHandler.java [hunk 4]
```diff
@@ -118,9 +123,9 @@ public void processReplyTo(MuleEvent event, MuleMessage returnMessage, Object re
...
+            String persistentDeliveryString = (String) eventMsg.getOutboundProperty(JmsConstants.PERSISTENT_DELIVERY_PROPERTY);
```
Comment (fsgonz) [inline transports/jms/src/main/java/org/mule/transport/jms/JmsReplyToHandler.java]: static import

File: transports/jms/src/main/java/org/mule/transport/jms/MultiConsumerJmsMessageReceiver.java [hunk 5]
```diff
@@ -235,7 +239,7 @@ protected void doDisconnect() throws Exception
         logger.debug("doDisconnect()");
 
         SubReceiver sub;
-        for (Iterator<SubReceiver> it = consumers.iterator(); it.hasNext();)
+        for (Iterator<SubReceiver> it = consumers.iterator(); it.hasNext(); )
```
Comment (fsgonz) [inline transports/jms/src/main/java/org/mule/transport/jms/MultiConsumerJmsMessageReceiver.java]: remove extra space.

File: transports/jms/src/main/java/org/mule/transport/jms/transformers/AbstractJmsTransformer.java [hunk 6]
```diff
@@ -124,7 +125,7 @@ protected Object transformFromMessage(Message source, String outputEncoding) thr
         if (logger.isDebugEnabled())
         {
             logger.debug("Message type received is: " +
-                    ClassUtils.getSimpleName(source.getClass()));
+                         ClassUtils.getSimpleName(source.getClass()));
```
Comment (fsgonz) [inline transports/jms/src/main/java/org/mule/transport/jms/transformers/AbstractJmsTransformer.java]: static import.

File: transports/jms/src/test/java/org/mule/transport/jms/DeferredCloserTestCase.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    private void notifyImClosingSomething(Object something)
+    {
+        System.out.printf("Something arrived to close #%d: %s\n", closureCounter.getAndIncrement(), something.toString());
+    }
+
+    private void probeGetsEmptied()
+    {
+        new PollingProber(10000, 1000).check(new Probe()
```
Comment (fsgonz) [inline transports/jms/src/test/java/org/mule/transport/jms/DeferredCloserTestCase.java]: use constants.

**Comment by plbalbi** [comment]
--validate

**Pull request merged**